### PR TITLE
[BugFix] Restrict partition column type of MV and only support date/fixpoint type

### LINKED ISSUE #19366: FE may restart with failure when create MV with varchar partition column
### Steps to reproduce the behavior (Required)
1. CREATE MV 
```
CREATE MATERIALIZED VIEW mv_db.tb_d
PARTITION BY date_str
DISTRIBUTED BY hash(log_id)
PROPERTIES (
"force_external_table_query_rewrite" = "true"
) 
AS
select log_id, get_json_string(json_string,'$.col') as col_id,create_time, date_str
from hudi_catalog0.db_ods.tb 
where date_str= '2023-03-10' and hour_str = '10'
```
`date_str` type is varchar
2. Refresh MV
3. Restart FE

### Expected behavior (Required)
FE restart successfully

### Real behavior (Required)
```
java.io.IOException: type[VARCHAR] not supported:
        at com.starrocks.catalog.PartitionKey.readFields(PartitionKey.java:364) ~[starrocks-fe.jar:?]
        at com.starrocks.catalog.PartitionKey.read(PartitionKey.java:374) ~[starrocks-fe.jar:?]
        at com.starrocks.common.util.RangeUtils.readRange(RangeUtils.java:166) ~[starrocks-fe.jar:?]
        at com.starrocks.persist.PartitionPersistInfo.readFields(PartitionPersistInfo.java:121) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.JournalEntity.readFields(JournalEntity.java:226) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.deserializeData(BDBJournalCursor.java:236) [starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.next(BDBJournalCursor.java:280) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournalInner(GlobalStateMgr.java:1894) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournal(GlobalStateMgr.java:1853) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.transferToLeader(GlobalStateMgr.java:1031) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.access$100(GlobalStateMgr.java:292) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$1.transferToLeader(GlobalStateMgr.java:644) [starrocks-fe.jar:?]
        at com.starrocks.ha.StateChangeExecutor.runOneCycle(StateChangeExecutor.java:86) [starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:115) [starrocks-fe.jar:?]
2023-03-10 14:58:04,244 WARN (stateChangeExecutor|81) [GlobalStateMgr.replayJournalInner():1913] catch exception when replaying 265966,
com.starrocks.journal.JournalException: fail to read journal entity key=265966, data=<DatabaseEntry offset="0" size="1368" data="0 12 0 0 0 0 0 0 44 91 0 0 0 0 0 0 87 29 255 255 255 255 255 255 255 255 255 255 255 255 255 255 255 255 0 0 0 0 0 0 87 32 0 0 0 9 112 50 48 50 50 49 48 50 48 0 0 0 6 78 79 82 77 65 76 255 255 255 255 255 255 255 255 255 255 255 255 255 255 255 255 0 0 0 0 0 0 87 30 0 0 0 6 78 79 82 77 65 76 0  ... 1268 bytes not shown "/>
        at com.starrocks.journal.bdbje.BDBJournalCursor.deserializeData(BDBJournalCursor.java:242) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.next(BDBJournalCursor.java:280) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournalInner(GlobalStateMgr.java:1894) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournal(GlobalStateMgr.java:1853) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.transferToLeader(GlobalStateMgr.java:1031) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.access$100(GlobalStateMgr.java:292) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$1.transferToLeader(GlobalStateMgr.java:644) [starrocks-fe.jar:?]
        at com.starrocks.ha.StateChangeExecutor.runOneCycle(StateChangeExecutor.java:86) [starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:115) [starrocks-fe.jar:?]
Caused by: java.io.IOException: type[VARCHAR] not supported:
        at com.starrocks.catalog.PartitionKey.readFields(PartitionKey.java:364) ~[starrocks-fe.jar:?]
        at com.starrocks.catalog.PartitionKey.read(PartitionKey.java:374) ~[starrocks-fe.jar:?]
        at com.starrocks.common.util.RangeUtils.readRange(RangeUtils.java:166) ~[starrocks-fe.jar:?]
        at com.starrocks.persist.PartitionPersistInfo.readFields(PartitionPersistInfo.java:121) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.JournalEntity.readFields(JournalEntity.java:226) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.deserializeData(BDBJournalCursor.java:236) ~[starrocks-fe.jar:?]
        ... 8 more
2023-03-10 14:58:04,245 ERROR (stateChangeExecutor|81) [GlobalStateMgr.transferToLeader():1036] failed to init journal after transfer to leader! will exit
com.starrocks.journal.JournalException: fail to read journal entity key=265966, data=<DatabaseEntry offset="0" size="1368" data="0 12 0 0 0 0 0 0 44 91 0 0 0 0 0 0 87 29 255 255 255 255 255 2568 bytes not shown "/>
        at com.starrocks.journal.bdbje.BDBJournalCursor.deserializeData(BDBJournalCursor.java:242) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.next(BDBJournalCursor.java:280) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournalInner(GlobalStateMgr.java:1894) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.replayJournal(GlobalStateMgr.java:1853) ~[starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.transferToLeader(GlobalStateMgr.java:1031) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr.access$100(GlobalStateMgr.java:292) [starrocks-fe.jar:?]
        at com.starrocks.server.GlobalStateMgr$1.transferToLeader(GlobalStateMgr.java:644) [starrocks-fe.jar:?]
        at com.starrocks.ha.StateChangeExecutor.runOneCycle(StateChangeExecutor.java:86) [starrocks-fe.jar:?]
        at com.starrocks.common.util.Daemon.run(Daemon.java:115) [starrocks-fe.jar:?]
Caused by: java.io.IOException: type[VARCHAR] not supported:
        at com.starrocks.catalog.PartitionKey.readFields(PartitionKey.java:364) ~[starrocks-fe.jar:?]
        at com.starrocks.catalog.PartitionKey.read(PartitionKey.java:374) ~[starrocks-fe.jar:?]
        at com.starrocks.common.util.RangeUtils.readRange(RangeUtils.java:166) ~[starrocks-fe.jar:?]
        at com.starrocks.persist.PartitionPersistInfo.readFields(PartitionPersistInfo.java:121) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.JournalEntity.readFields(JournalEntity.java:226) ~[starrocks-fe.jar:?]
        at com.starrocks.journal.bdbje.BDBJournalCursor.deserializeData(BDBJournalCursor.java:236) ~[starrocks-fe.jar:?]
```

### StarRocks version (Required)
 - You can get the StarRocks version by executing SQL `select current_version()`

### PR DESCRIPTION
**Pull request by caneGuy**
## What type of PR is this：
- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #19366 

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [x] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

File: fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java [hunk 1]
```diff
@@ -530,14 +531,15 @@ private void checkPartitionColumnWithBaseOlapTable(SlotRef slotRef, OlapTable ta
...
+            List<Column> partitionColumns = table.getPartitionColumns();
             if (table.isUnPartitioned()) {
                 throw new SemanticException("Materialized view partition column in partition exp " +
                         "must be base table partition column");
             } else {
                 boolean found = false;
-                for (String partitionColumn : partitionColumnNames) {
-                    if (partitionColumn.equalsIgnoreCase(slotRef.getColumnName())) {
+                for (Column partitionColumn : partitionColumns) {
+                    checkPartitionColumnType(partitionColumn);
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/analyzer/MaterializedViewAnalyzer.java:541]: checkPartitionColumnType after column name equals.

**Review by Astralidea — DISMISSED** [review]

**Comment by caneGuy** [comment]
run starrocks_fe_unittest

**Comment by caneGuy** [comment]
run starrocks_fe_unittest

**Review by Astralidea — DISMISSED** [review]

**Review by Astralidea — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-2.5

**Pull request merged**